### ISSUE TITLE
`serenity:all` re-exports triggers endless RAM-eating loop

### ISSUE THREAD

**Issue by Milo123459** [cargo-public-api/cargo-public-api#410]
I'm running `cargo public-api diff latest` on the next branch of this repository: https://github.com/serenity-rs/serenity on a github codespace with 32gb of ram. It keeps saying "Killed" after I run the command, suggesting it is OOMing. What can I do?

**Comment by Enselic** [comment]
I measured available system memory every 0.1 seconds while running that command on serenity (commit 6e2e70766e):

```
while true ; do sed -n 's/MemAvailable:   \([0-9]\+\) kB/\1/p' /proc/meminfo >> /tmp/mem-available.txt ; sleep 0.1 ; done
```

The plot looks like this, and the maximum dip in available memory while running that command is 900 MB.

![image](https://user-images.githubusercontent.com/115040/235413317-c823253c-8aa5-4f1b-b73b-396842eaed2c.png)

Running GNU time reports memory usage of the same order of magnitude, namely 680 MB:

```
$ \time --verbose cargo public-api diff latest
[...]
	Maximum resident set size (kbytes): 680016
```

So there should be no issue for your 32gb ram machine.

What numbers do you get with the methods above?

**Comment by Enselic** [comment]
Hmm hang on, I forgot to switch to the `next` branch in serenity (b9dadf1a907d). When I do that, reported memory usage becomes 29 GB by time and 18 GB in the graph. Something is fishy here...

**Comment by Enselic** [comment]
I did a git bisect, and cargo-public-api RAM usage increased from ~750 MB to ~32 000 MB with https://github.com/serenity-rs/serenity/commit/00f15f0edc648a28cb1e84377bfa0f681c6940bb, which began to re-export most public items from the crate. Intuitively I would expect around a doubling in RAM usage from such a change. Maybe there is some kind of bounded re-export recursion going on. We've had such problems before.

**Comment by Milo123459** [comment]
Is there a way to fix this on `cargo-public-api`s side?

**Comment by Enselic** [comment]
Yes, this seems to be a cargo-public-api bug. The bug seems to be caused by a glob import item importing another glob import item. I'm trying to figure out how to best fix it, and what a minimal reproducer (for a regression test) looks like.

Thanks a lot for reporting this issue btw.

**Comment by Milo123459** [comment]
Thank you for the quick response! Awesome project btw :)

**Comment by Enselic** [comment]
I'm confused as to why I thought RAM usage was capped at 32 GB. This was an endless RAM-eating loop as far as I can tell.

Should be fixed by #415

**Comment by Enselic** [comment]
Fix published in [0.29.1](https://crates.io/crates/cargo-public-api/0.29.1)
